runtime/interpreter/interpreter_expression.go

```diff
@@ -1335,6 +1335,12 @@ func (interpreter *Interpreter) VisitCastingExpression(expression *ast.CastingEx
 		// thus this is the only place where it becomes necessary to "instantiate" the result of a map to its
 		// concrete outputs. In other places (e.g. interface conformance checks) we want to leave maps generic,
 		// so we don't substitute them.
+
+		// if the target is anystruct or anyresource we want to preserve optionals
+		if !(expectedType == sema.AnyStructType || expectedType == sema.AnyResourceType) {
```

Comment:
We might have to also unwrap `expectedType` type before performing this check.
For e.g: If the targetType is an optional of AnyStruct (`AnyStruct?`), then it will pass the check and will unbox the optional value, which it shouldn't. i.e:
```cadence
let x: Int??? = 42
let y: AnyStruct = x as! AnyStruct?
```

Reply:
Good point, added this unwrapping.